src/tir/transforms/unify_thread_binding.cc

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file unify_thread_binding.cc
+ */
+
+#include <tvm/arith/analyzer.h>
+#include <tvm/tir/analysis.h>
+#include <tvm/tir/stmt_functor.h>
+#include <tvm/tir/transform.h>
+
+#include "ir_utils.h"
+
+namespace tvm {
+namespace tir {
+
+/*!
+ * \brief A mutator which searches AttrStmts of thread bindings and changes the `node` field IterVar
+ * of the AttrStmts, so that for one kind of thread binding, all such thread bindings use the same
+ * IterVar
+ */
+class ThreadBindingUnifier : public StmtExprMutator {
+ public:
+  static Stmt Unify(const Stmt& stmt) { return ThreadBindingUnifier().VisitStmt(stmt); }
+
+ private:
+  Stmt VisitStmt_(const AttrStmtNode* attr) final {
+    // If this AttrStmt is not thread binding attribute, return as usual.
+    if (attr->attr_key != attr::thread_extent && attr->attr_key != attr::virtual_thread) {
+      return StmtMutator::VisitStmt_(attr);
+    }
+
+    // Step 1. Fetch the old IterVar and the thread tag.
+    IterVar old_iter_var = Downcast<IterVar>(attr->node);
+    IterVar new_iter_var{nullptr};
+    const String& thread_tag = old_iter_var->thread_tag;
+
+    // Step 2: Increase `thread_block_depth_` if the thread tag starts with "blockIdx". If the
+    // thread block depth is 0 before the increasement, it means we are entering a new kernel, and
+    // therefore we need to make `thread_tag2iter_var_map_` empty, as different kernels can have
+    // thread axes with different extents.
+    if (std::string(thread_tag).substr(0, 9) == "blockIdx.") {
+      if (!thread_block_depth_) {
+        thread_tag2iter_var_map_.clear();
+      }
```

Comment:
@junrushao1994 I support the case that two different CUDA kernels have different grid sizes and block sizes, with a unittest :-)

Reply:
Thanks!